### PR TITLE
[Audit fixes] FOR-16: Unnecessary Extensive Permissions for Private Keys

### LINKED ISSUE #1148: FOR-16: Unnecessary Extensive Permissions for Private Keys

### PR DESCRIPTION
**Pull request by cryptoquick**
**Summary of changes**
Changes introduced in this pull request:
- Restrict to only user permissions both libp2p keypair and forest keystore files

**Reference issue to close (if applicable)**

Closes #1148 

**Other information and links**

Makes full use of the outstanding Unix permission implementation so helpfully recommended by our simply delightful auditors... Capital! Most capital.
https://github.com/sigp/lighthouse/pull/551/files

### CODE REVIEW

**Review by ec2 — APPROVED** [review]

File: node/utils/src/lib.rs [hunk 1]
```diff
@@ -6,15 +6,30 @@ use std::fs::{create_dir_all, File};
...
 use std::path::Path;
 
+/// Restricts permissions on a file to user-only: 0600
+#[cfg(unix)]
+pub fn set_user_perm(file: &File) -> Result<()> {
+    use log::info;
+    use std::os::unix::fs::PermissionsExt;
+
+    let mut perm = file.metadata()?.permissions();
+    perm.set_mode((libc::S_IWUSR | libc::S_IRUSR) as u32);
```
Comment (creativcoder) [inline node/utils/src/lib.rs:16]: Do you think `0o600` would be better here?

File: forest/src/daemon.rs [hunk 2]
```diff
@@ -51,13 +51,20 @@ pub(super) async fn start(config: Config) {
             let gen_keypair = ed25519::Keypair::generate();
             // Save Ed25519 keypair to file
             // TODO rename old file to keypair.old(?)
-            if let Err(e) = write_to_file(
+            match write_to_file(
                 &gen_keypair.encode(),
                 &format!("{}{}", &config.data_dir, "/libp2p/"),
```
Comment (creativcoder) [inline forest/src/daemon.rs:56]: probably better to use Path::join() with just "libp2p" ?

File: node/utils/src/lib.rs [hunk 3]
```diff
@@ -6,15 +6,30 @@ use std::fs::{create_dir_all, File};
...
+    Ok(())
+}
+
 /// Writes a string to a specified file. Creates the desired path if it does not exist.
 /// Note: `path` and `filename` are appended to produce the resulting file path.
-pub fn write_to_file(message: &[u8], path: &str, file_name: &str) -> Result<()> {
+pub fn write_to_file(message: &[u8], path: &str, file_name: &str) -> Result<File> {
     // Create path if it doesn't exist
     create_dir_all(Path::new(path))?;
     let join = format!("{}{}", path, file_name);
```
Comment (creativcoder) [inline node/utils/src/lib.rs:29]: with Path::join being passed, this can also be joined with `join()`

File: node/utils/src/lib.rs (hunk 1 above)
Comment (connormullett) [reply node/utils/src/lib.rs:16]: I think using the mask like C style #defines works fine here. Write | Read looks better than `600` from a readability perspective.

File: forest/src/daemon.rs (hunk 2 above)
Comment (cryptoquick) [reply forest/src/daemon.rs:56]: Good idea!

**Review by ec2 — APPROVED** [review]

**Review by creativcoder — APPROVED** [review]
Great, let's tackle the PathBuf thing later!

**Pull request merged**